container-images/assign-class-label/mutate.py

```diff
@@ -0,0 +1,180 @@
+
+
+def get_client() -> DynamicClient:
+    try:
+        config.load_config()
+        k8s_client = client.ApiClient()
+        dyn_client = DynamicClient(k8s_client)
+        return dyn_client
+    except config.ConfigException as e:
+        LOG.error('Could not configure Kubernetes client: %s', str(e))
+        exit(1)
+
+
+def get_group_resource(dyn_client):
+    return dyn_client.resources.get(api_version='user.openshift.io/v1', kind='Group')
+
+
+# Get users of a given group
+def get_group_members(group_resource: Any, group_name: str) -> List[str]:
+    group_obj = group_resource.get(name=group_name)
+    return group_obj.users
+
+
+def assign_class_label(
+    pod: Dict[str, Any], groups: List[str], dyn_client: DynamicClient
+) -> Optional[str]:
+    # Extract pod metadata
+    try:
+        pod_metadata = pod.get('metadata', {})
+        pod_labels = pod_metadata.get('labels', {})
+        pod_user = pod_labels.get('opendatahub.io/user', None)
+    except AttributeError as e:
+        LOG.error(f'Error extracting pod information: {e}')
+        return None
+    
+    if pod_user is not None:
+        pod_user = decode_pod_user(pod_user)
+
+    group_resource = get_group_resource(dyn_client)
+
+    # Iterate through classes
+    for group in groups:
+        users = get_group_members(group_resource, group)
+
+        # Check if group has no users
+        if not users:
+            LOG.warning(f'Group {group} has no users or users attribute is not a list.')
+            continue
+
+        # Compare users in the groups (classes) with the pod user
+        if pod_user in users:
+            LOG.info(f'Assigning class label: {group} to user {pod_user}')
+            return group
+
+    return None
+
+
+def create_app(**config: Any) -> Flask:
+    app = Flask(__name__)
+    app.config.from_prefixed_env('RHOAI_CLASS')
+    app.config.update(config)
+
+    if not app.config['GROUPS']:
+        LOG.error('RHOAI_CLASS_GROUPS environment variables are required.')
+        exit(1)
+
+    groups = app.config['GROUPS'].split(',')
+
+    dyn_client = get_client()
+
+    @app.route('/mutate', methods=['POST'])
+    def mutate_pod() -> Response:
+        # Grab pod for mutation and validate request
+        try:
+            admission_review = AdmissionReview(**request.get_json())
+        except ValidationError as e:
+            LOG.error('Validation error: %s', e)
+            return (
+                jsonify(
+                    {
+                        'apiVersion': 'admission.k8s.io/v1',
+                        'kind': 'AdmissionReview',
+                        'response': {
+                            'uid': request.json.get('request', {}).get('uid', ''),
+                            'allowed': False,
+                            'status': {'message': f'Invalid request: {e}'},
+                        },
+                    }
+                ),
+                400,
+                {'content-type': 'application/json'},
+            )
```

Comment:
You should be using a pydantic model to generate this response, rather than manually constructing the response structure. This provides the same validation for your output that you get when reading input in the previous step.